library/cwm/src/lib/cwm/ui_state.rb

```diff
@@ -0,0 +1,209 @@
+  #
+  #   require "cwm/ui_state"
+  #   class MyUIState < CWM::UIState
+  #     def go_to_tree_node(page)
+  #       super
+  #       self.candidate_notes =
+  #         if page.respond_to?(:zone)
+  #           zone_page_candidates(page)
+  #         else
+  #           [page.label]
+  #         end
+  #       end
+  #     end
+  #
+  #   private
+  #
+  #     def zone_page_candidates(page)
+  #       [page.zone.name]
+  #     end
+  #   end
+  #
+  # @example Using the UIState from an CWM::TreePager
+  #
+  #   class OverviewTreePager < CWM::TreePager
+  #     # Overrides default behaviour to register the new state.
+  #     def switch_page(page)
+  #       MyUIState.instance.go_to_tree_node(page)
+  #       super
+  #     end
+  #
+  #     # Ensures the tree is properly initialized according to the UI state after
+  #     # a redraw
+  #     def initial_page
+  #       MyUIState.instance.find_tree_node(@pages) || super
+  #     end
+  #   end
+  #
+  # @example Registering the selected row from a CWM::Table
+  #
+  #   class MyTable < ::CWM::Table
+  #     def opt
+  #       [:notify, :immediate]
+  #     end
+  #
+  #     def handle(event)
+  #       MyUIState.instance.select_row(value) if event["EventReason"] == "SelectionChanged"
+  #       nil
+  #     end
+  #   end
+  #
+  class UIState
+    include Yast::I18n
+
+    # Constructor
+    #
+    # Called through {.create_instance}, starts with a blank situation (which
+    # means default for each widget will be honored).
+    def initialize
+      textdomain(textdomain_name)
+      @candidate_nodes = []
+    end
+
+    # Specify textdomain name
+    #
+    # @return [String] Textdomain (e.g., "firewall")
+    abstract_method :textdomain_name
+
+    # Method to be called when the user decides to visit a given page by
+    # clicking in one node of the general tree.
+    #
+    # It remembers the decision so the user is taken back to a sensible point of
+    # the tree (very often the last he decided to visit) after redrawing.
+    #
+    # @param [CWM::Page] page associated to the tree node
+    def go_to_tree_node(page)
+      self.candidate_nodes = [page.label]
+
+      # Landing in a new node, so invalidate previous details about position
+      # within a node, they no longer apply
+      self.tab = nil
+    end
+
+    # Method to be called when the user switches to a tab within a tree node.
+    #
+    # It remembers the decision so the same tab is showed in case the user stays
+    # in the same node after redrawing.
+    #
+    # @param [CWM::Page] page associated to the tab
+    def switch_to_tab(page)
+      self.tab = page.label
+    end
+
+    # Method to be called when the user operates in a row of a table.
+    #
+    # @param row_id [Object] row identifier
+    def select_row(row_id)
+      self.row_id = row_id
+    end
+
+    # Select the page to open in the general tree after a redraw
```

Comment:
It is strange to mention select when in fact you do find.

Reply:
Yes, it sounds confusing. But I decided to keep the original API as much as possible.